### ISSUE TITLE
`net-snmp` based validation testing

### ISSUE THREAD

**Issue by upsampled** [gosnmp/gosnmp#448]
Just as `crypto/tls` uses `openssl` to validate their implementation (although in a complicated record and playback scheme), I think `gosnmp` would benefit from validation testing against `net-snmp`.

Here is what I was planning initially:
* test files that do this would be named `*_netsnmp_test.go` and have the build tag `//go:build netsnmp`
* they will use `cgo` to link against `libsnmp` and it is assumed to be installed locally 
* they will first probe `libsnmp` to ensure it has the feature being tested (IE: check that your cipher suite is there)
* generate the expected value from `libanmp` then run `gosnmp/gosnmp` to generate the gotten value (`if gen != got {....}`)

I was panning on just making sure PDU encoding works first in the initial PR (also test `Guage32` and add `Counter64` support), but the framework can be extended to other parts in the future.

Does this sound like an acceptable validation framework?

**Comment by upsampled** [comment]
I am dumb and forgot `cgo` cannot directly be in testing. I am going to shim around that and come up with a proof-of-concept and you can decide if I should proceed or not.

**Comment by TimRots** [comment]
> ..
> Does this sound like an acceptable validation framework?

I personally like the idea of doing more validation against net-snmp. Not sure yet if I would be a proponent of introducing cgo bindings to interface with libsnmp. Interested to see what PoC you come up with.

**Comment by upsampled** [comment]
Well that was a pain, but have something that is mostly working. Here is my recomended path forward:
1.  Get the Proof-of-concept entirely working
2.  Determine if/how the final validation testing framework should be implemented
3.  Implement it for a very scoped section of code (aiming for PDU encoding)
4.  Use other tickets to track adoption to other sections of code

So regarding step 1, 
* Keeping the proof-of-concept code here: https://github.com/upsampled/gosnmp/tree/netsnmp-proof-of-concept
* Assmune you have `libsnmp-dev` installed, you can run the test with: `go test -tags=netsnmp -v -run TestTypeGauge32PDU`
* It is currently failing with:
```
=== RUN   TestTypeGauge32PDU
    gosnmp_netsnmp_test.go:40: got(42): "0(\x02\x01\x01\x04\x06public\xa3\x1b\x02\x04\x00\x00\x00\x00\x02\x01\x00\x02\x01\x000\r0\v\x06\x05*\x03\a\t\nB\x02\x04\xd7"
        exp(47): "0\x82\x00+\x02\x01\x01\x04\x06public\xa3\x82\x00\x1c\x02\x01\x01\x02\x01\x00\x02\x01\x000\x82\x00\x0f0\x82\x00\v\x06\x05*\x03\a\t\nB\x02\x04\xd7"
        diff:
          []uint8{
                0x30,
        -       0x82,
        +       0x28,
        -       0x00,
        -       0x2b,
                0x02,
                0x01,
                ... // 8 identical elements
                0x63,
                0xa3,
        -       0x82,
        +       0x1b,
        -       0x00,
        -       0x1c,
                0x02,
        -       0x01,
        +       0x04,
        -       0x01,
        +       0x00,
        +       0x00,
        +       0x00,
        +       0x00,
                0x02,
                0x01,
                0x00,
                0x02,
                0x01,
        -       0x00,
        -       0x30,
        -       0x82,
                0x00,
        +       0x30,
        -       0x0f,
        +       0x0d,
                0x30,
        -       0x82,
        -       0x00,
                0x0b,
                0x06,
                ... // 10 identical elements
          }
--- FAIL: TestTypeGauge32PDU (0.00s)
```

Generally, it looks like they are close, but i need to manually decrypt packets to figure out where they differ. If you have insight that be great, if not then I'll get around to it eventually.

**Comment by upsampled** [comment]
@TimRots I think this testing is already finding interesting differences. It looks like both the packets generated are valid, and the difference is the `net-snmp` is using the [long form](https://www.oss.com/asn1/resources/asn1-made-simple/asn1-quick-reference/basic-encoding-rules.html)(`0x82`) of encoding ASN.1 header lengths for the individual sections. Anyhow I'll see if I can turn that off in `net-snmp` or turn on long form in the gosnmp code just for testing.

**Comment by upsampled** [comment]
@TimRots and @SuperQ  thank you! This unblocks the `GoSMPServer` migration so there is now a unified Go SNMP Server implementation.

Anyhow, as far as future work, I was thinking the following:
- Using the validation framework with fuzzing should be extremely beneficial. I only caught #453 because of a test in GoSNMPServer's suite. 
- Move more encoding over to `asn1`. I think #451 pointed out that the `ans.1` package better suites to encode PDUs. The only thing it really is lacking is `Opaque*` types (which i found out are only based on a[ draft RFC](https://datatracker.ietf.org/doc/html/draft-perkins-opaque-01)). Anyhow I opened up a ticket golang/go#63399 to see if the support can be added. Figure I'll wait to see how that resolves before making any additional PRs.
- More of a nit, but it might be beneficial to wrap the crypto part of the USM into a `cipher.Stream` and it would assume every  `XORKeyStream` is a full packet. Would be a cleaner way to handle #447.
